test-common/src/main/scala/org/scalajs/testcommon/RPCCore.scala

```diff
@@ -102,19 +114,28 @@ private[scalajs] abstract class RPCCore {
     // Reserve an id for this call.
     val id = nextID.incrementAndGet()
 
-    // Prepare message.
+    // Prepare message. We do this early in case it throws.
     val msg = makeRPCMsg(opCode, id, req)
 
     // Register pending call.
     val promise = Promise[Resp]
     val oldCall = pending.put(id, PendingCall(promise))
 
     if (oldCall != null) {
-      throw new AssertionError("Ran out of call ids!")
+      val error = new AssertionError("Ran out of call ids!")
+      close(error)
+      throw error
     }
 
-    // Actually send message.
-    send(msg)
+    if (closeReason != null) {
+      /* In the meantime, someone closed the channel. Help closing.
+       * We need this check to guard against a race between `call` and `close`.
+       */
+      helpClose()
+    } else {
+      // Actually send message.
+      send(msg)
```

Comment:
What happens if `closeReason` is set to non-null between the `if (closeReason != null)` and the `send(msg)`?

Reply:
Then whatever just set `closeReason` will continue to call `helpClose`. There it will see this invocation in `pending` and race with the server to either fail or succeed it. In both cases, the pending call will not be left dangling.

Reply:
OK thanks.